fontc/src/main.rs

```diff
@@ -255,8 +247,139 @@ impl ChangeDetector {
+            })
+            .collect()
+    }
+    fn exec(mut self, scope: &Scope, root_context: Context) -> Result<JobResults, Error> {
+        let job_count = self.jobs.len();
+
+        // Async work will send us it's ID on completion
+        let (send, recv) =
+            crossbeam_channel::unbounded::<(WorkIdentifier, Result<(), WorkError>)>();
+        let mut success = HashSet::new();
+        let mut error = HashSet::new();
+
+        // Whenever a task completes see if it was the last incomplete dependency of other task(s)
+        // and spawn them if it was
+        // TODO timeout and die it if takes too long to make forward progress or we're spinning w/o progress
+        while success.len() < job_count && error.is_empty() {
+            // Spawn anything that is currently executable (has no unfulfilled dependencies)
+            for id in self.launchable(&success) {
+                trace!("Start {:?}", id);
+                let job = self.jobs.remove(&id).unwrap();
+                let work = job.work;
+                let work_context = root_context.copy_for_work(id.clone(), Some(job.happens_after));
+                let send = send.clone();
+                scope.spawn(move |_| {
+                    let result = work.exec(&work_context);
+                    if let Err(e) = send.send((id.clone(), result)) {
+                        error!("Unable to write {:?} to completion channel: {}", id, e);
+                    }
+                })
+            }
+
+            // Block for things to phone home to say they are done
+            // Then complete everything that has reported since our last check
+            let mut opt_complete = Some(recv.recv().unwrap()); // blocks
+            while let Some((completed_id, result)) = opt_complete.take() {
+                if !match result {
+                    Ok(..) => success.insert(completed_id.clone()),
+                    Err(..) => error.insert(completed_id.clone()),
+                } {
+                    panic!("Repeat signals for completion of {:#?}", completed_id);
+                }
+                debug!(
+                    "{}/{} complete, most recently {:?}",
+                    error.len() + success.len(),
+                    job_count,
+                    completed_id
+                );
+
+                // See if anything else is complete in case things come in waves
+                if let Ok(completed_id) = recv.try_recv() {
+                    opt_complete = Some(completed_id);
+                }
+            }
+        }
+        if success.len() != job_count {
+            panic!("No errors but not everything succeeded?!");
+        }
+        Ok(JobResults { success, error })
+    }
+}
+
+/// A unit of executable work plus the identifiers of work that it depends on
+struct Job {
+    // The actual task
+    work: Box<dyn Work + Send>,
+    // Things that must happen before us, e.g. our dependencies
+    happens_after: HashSet<WorkIdentifier>,
```

Comment:
I think `dependencies` would be better/clearer name for this field than `happens_after`.. The latter being a verb instead of a noun, it evokes to me an action more than a state, thus more a method name than the name of an attribute..

Reply:
For context happens-after is more how you would think about it when reasoning about the concurrency. I think I agree re: renaming though.

Reply:
https://github.com/googlefonts/fontmake-rs/pull/77